projects/fal/src/fal/app.py

```diff
@@ -170,7 +181,176 @@ def marker_fn(callable: EndpointT) -> EndpointT:
+                    websocket.receive_bytes(),
+                    timeout=session_timeout,
+                )
+            except asyncio.TimeoutError:
+                return
+
+            input = msgpack.unpackb(raw_input, raw=False)
+            if input_modal:
+                input = input_modal(**input)
+
+            queue.append(input)
+
+    async def mirror_output(self, websocket: WebSocket) -> None:
+        while True:
+            if not queue:
+                await asyncio.sleep(0.05)
+                continue
+
+            input = queue.popleft()
+            if input is None:
+                break
+
+            output = func(self, input)
+            if not isinstance(output, dict):
+                # Handle pydantic output modal
+                if hasattr(output, "dict"):
+                    output = output.dict()
+                else:
+                    raise TypeError(
+                        f"Expected a dict or pydantic model as output, got {type(output)}"
+                    )
+
+            message = msgpack.packb(output, use_bin_type=True)
+            await websocket.send_bytes(message)
+
+    async def websocket_template(self, websocket: WebSocket) -> None:
+        import asyncio
+
+        await websocket.accept()
+        input_task = asyncio.create_task(mirror_input(websocket))
+        input_task.add_done_callback(lambda _: queue.append(None))
+        output_task = asyncio.create_task(mirror_output(self, websocket))
+        try:
+            await asyncio.wait(
+                {
+                    input_task,
+                    output_task,
+                },
+                return_when=asyncio.FIRST_COMPLETED,
+            )
+            if input_task.done():
+                # User didn't send any input within the timeout
+                # so we can just close the connection after the
+                # processing of the last input is done.
+                input_task.result()
+                await asyncio.wait_for(output_task, timeout=session_timeout)
+            else:
+                assert output_task.done()
+
+                # The execution of the inference function failed or exitted,
+                # so just propagate the result.
+                input_task.cancel()
+                with suppress(asyncio.CancelledError):
+                    await input_task
+
+                output_task.result()
+        except WebSocketDisconnect:
+            input_task.cancel()
+            output_task.cancel()
+            with suppress(asyncio.CancelledError):
+                await input_task
+
+            with suppress(asyncio.CancelledError):
+                await output_task
+        except Exception as exc:
+            import traceback
+
+            traceback.print_exc()
+
+            await websocket.send_json(
+                {
+                    "type": "x-fal-error",
+                    "error": "INTERNAL_ERROR",
+                    "reason": str(exc),
+                }
```

Comment:
how does the consumer determine whether the response is a `msgpack` struct or a `json`? or why don't we use `msgpack` for all the messages?

Reply:
fal emits its own messages as json, so this is done to mimic that. binary protocol is used only for I/O, and errors and other stuff goes as json. the users are expected to decode the frames to see if they are a text frame or a binary frame, and use the right format for that (json and msgpack respectively). https://github.com/fal-ai/fal-js/blob/12887d4b876793ef060924e5bc10a08d39d6313a/libs/client/src/realtime.ts#L473-L495